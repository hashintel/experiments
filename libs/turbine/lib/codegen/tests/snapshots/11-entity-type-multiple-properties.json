[
  {
    "$id": "http://localhost:3000/@alice/types/entity-type/country/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
    "allOf": [],
    "description": null,
    "kind": "entityType",
    "links": {},
    "properties": {
      "http://localhost:3000/@alice/types/property-type/countryCode/": {
        "$ref": "http://localhost:3000/@alice/types/property-type/countryCode/v/1"
      },
      "http://localhost:3000/@alice/types/property-type/name/": {
        "$ref": "http://localhost:3000/@alice/types/property-type/name/v/1"
      }
    },
    "required": [
      "http://localhost:3000/@alice/types/property-type/countryCode/",
      "http://localhost:3000/@alice/types/property-type/name/"
    ],
    "title": "Country",
    "type": "object"
  },
  {
    "$id": "http://localhost:3000/@alice/types/property-type/name/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
    "description": null,
    "kind": "propertyType",
    "oneOf": [
      {
        "$ref": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1"
      }
    ],
    "title": "Name"
  },
  {
    "$id": "http://localhost:3000/@alice/types/property-type/countryCode/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
    "description": "(ISO 3166-1 alpha-2)",
    "kind": "propertyType",
    "oneOf": [
      {
        "$ref": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1"
      }
    ],
    "title": "Country Code"
  },
  {
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    "kind": "dataType",
    "$id": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1",
    "title": "Text",
    "description": "An ordered sequence of characters",
    "type": "string"
  }
]
