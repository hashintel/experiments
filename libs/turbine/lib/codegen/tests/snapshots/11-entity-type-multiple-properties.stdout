localhost_3000/alice/entity/country.rs

use alloc::string::String;

use error_stack::{Report, Result, ResultExt as _};
use hashbrown::HashMap;
use serde::Serialize;
use turbine::{
    entity::Entity, url, BaseUrl, EntityLink, EntityProperties, EntityType, EntityTypeMut,
    EntityTypeRef, GenericEntityError, OptionalEntityLink, PropertyType as _, PropertyTypeMut as _,
    PropertyTypeRef as _, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};

use crate::localhost_3000::alice::property::{country_code::CountryCode, name::Name};
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Properties {
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/countryCode/")]
    pub country_code: CountryCode,
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/name/")]
    pub name: Name,
}
impl Properties {
    fn try_from_value(
        mut properties: HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        let country_code = 'property: {
            let value =
                properties.remove("http://localhost:3000/@alice/types/property-type/countryCode/");
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/countryCode/",
                )));
            };
            let value =
                <CountryCode>::try_from_value(value).change_context(GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/countryCode/",
                ));
            value
        };
        let name = 'property: {
            let value = properties.remove("http://localhost:3000/@alice/types/property-type/name/");
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/name/",
                )));
            };
            let value = <Name>::try_from_value(value).change_context(GenericEntityError::Property(
                "http://localhost:3000/@alice/types/property-type/name/",
            ));
            value
        };
        let __report0 = turbine::fold_tuple_reports((country_code, name));
        let ((country_code, name),) = turbine::fold_tuple_reports((__report0,))?;
        let this = Self { country_code, name };
        Ok(this)
    }

    fn as_mut(&mut self) -> PropertiesMut<'_> {
        let Self { country_code, name } = self;
        PropertiesMut {
            country_code: <CountryCode as Type>::as_mut(country_code),
            name: <Name as Type>::as_mut(name),
        }
    }

    fn as_ref(&self) -> PropertiesRef<'_> {
        let Self { country_code, name } = self;
        PropertiesRef {
            country_code: <CountryCode as Type>::as_ref(country_code),
            name: <Name as Type>::as_ref(name),
        }
    }
}
///Country
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Country {
    pub properties: Properties,
}
pub type CountryV1 = Country;
impl TypeUrl for Country {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl Type for Country {
    type Mut < 'a > = CountryMut < 'a > where Self : 'a ;
    type Ref < 'a > = CountryRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> Self::Mut<'_> {
        CountryMut {
            properties: self.properties.as_mut(),
        }
    }

    fn as_ref(&self) -> Self::Ref<'_> {
        CountryRef {
            properties: self.properties.as_ref(),
        }
    }
}
impl EntityType for Country {
    type Error = GenericEntityError;

    fn try_from_entity(value: Entity) -> Option<Result<Self, Self::Error>> {
        if Self::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = Properties::try_from_value(value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl EntityProperties for Country {
    type Properties = Properties;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for Country {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct PropertiesRef<'a> {
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/countryCode/")]
    pub country_code: <CountryCode as Type>::Ref<'a>,
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/name/")]
    pub name: <Name as Type>::Ref<'a>,
}
impl<'a> PropertiesRef<'a> {
    fn try_from_value(
        properties: &'a HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        let country_code = 'property: {
            let value =
                properties.get("http://localhost:3000/@alice/types/property-type/countryCode/");
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/countryCode/",
                )));
            };
            let value = <<CountryCode as Type>::Ref<'a>>::try_from_value(value).change_context(
                GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/countryCode/",
                ),
            );
            value
        };
        let name = 'property: {
            let value = properties.get("http://localhost:3000/@alice/types/property-type/name/");
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/name/",
                )));
            };
            let value = <<Name as Type>::Ref<'a>>::try_from_value(value).change_context(
                GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/name/",
                ),
            );
            value
        };
        let __report0 = turbine::fold_tuple_reports((country_code, name));
        let ((country_code, name),) = turbine::fold_tuple_reports((__report0,))?;
        let this = Self { country_code, name };
        Ok(this)
    }

    fn into_owned(self) -> Properties {
        let Self { country_code, name } = self;
        Properties {
            country_code: <<CountryCode as Type>::Ref<'_> as TypeRef>::into_owned(country_code),
            name: <<Name as Type>::Ref<'_> as TypeRef>::into_owned(name),
        }
    }
}
///Country
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CountryRef<'a> {
    pub properties: PropertiesRef<'a>,
}
pub type CountryV1Ref<'a> = CountryRef<'a>;
impl TypeUrl for CountryRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl TypeRef for CountryRef<'_> {
    type Owned = Country;

    fn into_owned(self) -> Self::Owned {
        Country {
            properties: self.properties.into_owned(),
        }
    }
}
impl<'a> EntityTypeRef<'a> for CountryRef<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesRef::try_from_value(&value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl<'a> EntityProperties for CountryRef<'a> {
    type Properties = PropertiesRef<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for CountryRef<'_> {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
#[derive(Debug, Serialize)]
pub struct PropertiesMut<'a> {
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/countryCode/")]
    pub country_code: <CountryCode as Type>::Mut<'a>,
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/name/")]
    pub name: <Name as Type>::Mut<'a>,
}
impl<'a> PropertiesMut<'a> {
    fn try_from_value(
        properties: &'a mut HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        let country_code = 'property: {
            let value = unsafe {
                let value = properties
                    .get_mut("http://localhost:3000/@alice/types/property-type/countryCode/");
                let value = value.map(|value| value as *mut _);
                value.map(|value: *mut serde_json::Value| &mut *value)
            };
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/countryCode/",
                )));
            };
            let value = <<CountryCode as Type>::Mut<'a>>::try_from_value(value).change_context(
                GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/countryCode/",
                ),
            );
            value
        };
        let name = 'property: {
            let value = unsafe {
                let value =
                    properties.get_mut("http://localhost:3000/@alice/types/property-type/name/");
                let value = value.map(|value| value as *mut _);
                value.map(|value: *mut serde_json::Value| &mut *value)
            };
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/name/",
                )));
            };
            let value = <<Name as Type>::Mut<'a>>::try_from_value(value).change_context(
                GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/name/",
                ),
            );
            value
        };
        let __report0 = turbine::fold_tuple_reports((country_code, name));
        let ((country_code, name),) = turbine::fold_tuple_reports((__report0,))?;
        let this = Self { country_code, name };
        Ok(this)
    }

    fn into_owned(self) -> Properties {
        let Self { country_code, name } = self;
        Properties {
            country_code: <<CountryCode as Type>::Mut<'_> as TypeMut>::into_owned(country_code),
            name: <<Name as Type>::Mut<'_> as TypeMut>::into_owned(name),
        }
    }
}
///Country
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CountryMut<'a> {
    pub properties: PropertiesMut<'a>,
}
pub type CountryV1Mut<'a> = CountryMut<'a>;
impl TypeUrl for CountryMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl TypeMut for CountryMut<'_> {
    type Owned = Country;

    fn into_owned(self) -> Self::Owned {
        Country {
            properties: self.properties.into_owned(),
        }
    }
}
impl<'a> EntityTypeMut<'a> for CountryMut<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a mut Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesMut::try_from_value(&mut value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl<'a> EntityProperties for CountryMut<'a> {
    type Properties = PropertiesMut<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for CountryMut<'_> {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}


---

localhost_3000/alice/property/country_code.rs

use error_stack::{Report, Result, ResultExt as _};
use serde::Serialize;
use turbine::{
    types::data::Text, url, DataType, DataTypeMut, DataTypeRef, GenericPropertyError, PropertyType,
    PropertyTypeMut, PropertyTypeRef, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};
///Country Code
///
///(ISO 3166-1 alpha-2)
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct CountryCode(pub Text);
impl TypeUrl for CountryCode {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/countryCode/" / v / 1u32);
}
impl Type for CountryCode {
    type Mut < 'a > = CountryCodeMut < 'a > where Self : 'a ;
    type Ref < 'a > = CountryCodeRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> CountryCodeMut<'_> {
        let Self(value) = self;
        CountryCodeMut(<Text as Type>::as_mut(value))
    }

    fn as_ref(&self) -> CountryCodeRef<'_> {
        let Self(value) = self;
        CountryCodeRef(<Text as Type>::as_ref(value))
    }
}
impl PropertyType for CountryCode {
    type Error = GenericPropertyError;

    fn try_from_value(value: serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <Text as DataType>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type CountryCodeV1 = CountryCode;
///Country Code
///
///(ISO 3166-1 alpha-2)
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct CountryCodeRef<'a>(pub <Text as Type>::Ref<'a>);
impl TypeUrl for CountryCodeRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/countryCode/" / v / 1u32);
}
impl TypeRef for CountryCodeRef<'_> {
    type Owned = CountryCode;

    fn into_owned(self) -> CountryCode {
        let Self(value) = self;
        CountryCode(<<Text as Type>::Ref<'_> as TypeRef>::into_owned(value))
    }
}
impl<'a> PropertyTypeRef<'a> for CountryCodeRef<'a> {
    type Error = GenericPropertyError;

    fn try_from_value(value: &'a serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <<Text as Type>::Ref<'a> as DataTypeRef<'a>>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type CountryCodeV1Ref<'a> = CountryCodeRef<'a>;
///Country Code
///
///(ISO 3166-1 alpha-2)
#[derive(Debug, Serialize)]
pub struct CountryCodeMut<'a>(pub <Text as Type>::Mut<'a>);
impl TypeUrl for CountryCodeMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/countryCode/" / v / 1u32);
}
impl TypeMut for CountryCodeMut<'_> {
    type Owned = CountryCode;

    fn into_owned(self) -> CountryCode {
        let Self(value) = self;
        CountryCode(<<Text as Type>::Mut<'_> as TypeMut>::into_owned(value))
    }
}
impl<'a> PropertyTypeMut<'a> for CountryCodeMut<'a> {
    type Error = GenericPropertyError;

    fn try_from_value(value: &'a mut serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <<Text as Type>::Mut<'a> as DataTypeMut<'a>>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type CountryCodeV1Mut<'a> = CountryCodeMut<'a>;


---

localhost_3000/alice/property/name.rs

use error_stack::{Report, Result, ResultExt as _};
use serde::Serialize;
use turbine::{
    types::data::Text, url, DataType, DataTypeMut, DataTypeRef, GenericPropertyError, PropertyType,
    PropertyTypeMut, PropertyTypeRef, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};
///Name
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct Name(pub Text);
impl TypeUrl for Name {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/name/" / v / 1u32);
}
impl Type for Name {
    type Mut < 'a > = NameMut < 'a > where Self : 'a ;
    type Ref < 'a > = NameRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> NameMut<'_> {
        let Self(value) = self;
        NameMut(<Text as Type>::as_mut(value))
    }

    fn as_ref(&self) -> NameRef<'_> {
        let Self(value) = self;
        NameRef(<Text as Type>::as_ref(value))
    }
}
impl PropertyType for Name {
    type Error = GenericPropertyError;

    fn try_from_value(value: serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <Text as DataType>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type NameV1 = Name;
///Name
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct NameRef<'a>(pub <Text as Type>::Ref<'a>);
impl TypeUrl for NameRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/name/" / v / 1u32);
}
impl TypeRef for NameRef<'_> {
    type Owned = Name;

    fn into_owned(self) -> Name {
        let Self(value) = self;
        Name(<<Text as Type>::Ref<'_> as TypeRef>::into_owned(value))
    }
}
impl<'a> PropertyTypeRef<'a> for NameRef<'a> {
    type Error = GenericPropertyError;

    fn try_from_value(value: &'a serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <<Text as Type>::Ref<'a> as DataTypeRef<'a>>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type NameV1Ref<'a> = NameRef<'a>;
///Name
#[derive(Debug, Serialize)]
pub struct NameMut<'a>(pub <Text as Type>::Mut<'a>);
impl TypeUrl for NameMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/name/" / v / 1u32);
}
impl TypeMut for NameMut<'_> {
    type Owned = Name;

    fn into_owned(self) -> Name {
        let Self(value) = self;
        Name(<<Text as Type>::Mut<'_> as TypeMut>::into_owned(value))
    }
}
impl<'a> PropertyTypeMut<'a> for NameMut<'a> {
    type Error = GenericPropertyError;

    fn try_from_value(value: &'a mut serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <<Text as Type>::Mut<'a> as DataTypeMut<'a>>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type NameV1Mut<'a> = NameMut<'a>;
